From 0d9fbcfb24067aac415700614666c75f2decdf47 Mon Sep 17 00:00:00 2001
From: Mike Drob <mdrob@cloudera.com>
Date: Thu, 5 Feb 2015 12:16:08 -0600
Subject: [PATCH] LUCENE-3973 Add lint task

Added lint task and a PMD task that calls it. Leaves room for future
expansion to other static analisys tools in lint.
---
 build.xml                 | 10 ++++++++++
 lucene/analysis/build.xml |  4 ++++
 lucene/build.xml          | 11 ++++++++++-
 lucene/common-build.xml   | 23 +++++++++++++++++++++++
 lucene/lucene-ruleset.xml | 46 ++++++++++++++++++++++++++++++++++++++++++++++
 solr/build.xml            |  9 +++++++++
 6 files changed, 102 insertions(+), 1 deletion(-)
 create mode 100644 lucene/lucene-ruleset.xml

diff --git build.xml build.xml
index 5d7935e..0ab306f 100644
--- build.xml
+++ build.xml
@@ -128,6 +128,16 @@
     </subant>
   </target>
 
+  <target name="lint" description="Runs static analysis across all sources"
+    depends="pmd" />
+  
+  <target name="pmd" description="Runs PMD across all sources">
+    <sequential><subant target="pmd" inheritall="false" failonerror="true">
+        <fileset dir="lucene" includes="build.xml" />
+        <fileset dir="solr" includes="build.xml" />
+    </subant></sequential>
+  </target>
+
   <target name="resolve" description="Resolves all dependencies">
     <subant target="resolve" inheritall="false" failonerror="true">
       <fileset dir="lucene" includes="build.xml" />
diff --git lucene/analysis/build.xml lucene/analysis/build.xml
index 0abd60a..d8cbd77 100644
--- lucene/analysis/build.xml
+++ lucene/analysis/build.xml
@@ -149,6 +149,10 @@
   <target name="-append-module-dependencies-properties">
     <forall-analyzers target="-append-module-dependencies-properties"/>
   </target>
+
+  <target name="pmd">
+    <forall-analyzers target="pmd"/>
+  </target>
 	
   <target name="check-forbidden-apis">
     <forall-analyzers target="check-forbidden-apis"/>
diff --git lucene/build.xml lucene/build.xml
index c1e0971..eca14dd 100644
--- lucene/build.xml
+++ lucene/build.xml
@@ -28,7 +28,7 @@
 
   <patternset id="binary.build.dist.patterns"
               includes="docs/,**/*.jar,**/*.war"
-              excludes="poms/**,**/*-src.jar,**/*-javadoc.jar"
+              excludes="pmd/**,poms/**,**/*-src.jar,**/*-javadoc.jar"
   />
   <patternset id="binary.root.dist.patterns"
               includes="LICENSE.txt,NOTICE.txt,README.txt,
@@ -250,6 +250,15 @@
     <modules-crawl target="rat-sources"/>
   </target>
 
+  <target name="pmd">
+     <sequential>
+       <ant dir="core" target="pmd" inheritall="false"/>
+       <ant dir="test-framework" target="pmd" inheritall="false"/>
+       <ant dir="tools" target="pmd" inheritall="false"/>
+       <modules-crawl target="pmd" failonerror="true"/>
+    </sequential>
+  </target>
+
   <!-- ================================================================== -->
   <!-- D I S T R I B U T I O N                                            -->
   <!-- ================================================================== -->
diff --git lucene/common-build.xml lucene/common-build.xml
index 478ead5..7cb686c 100644
--- lucene/common-build.xml
+++ lucene/common-build.xml
@@ -1834,6 +1834,29 @@ ${ant.project.name}.test.dependencies=${test.classpath.list}
     </fail>
   </target>
 
+  <target name="pmd-typedef" depends="ivy-availability-check,ivy-fail">
+    <ivy:cachepath pathid="pmd.classpath" organisation="org.apache.maven.plugins" module="maven-pmd-plugin" revision="3.4" inline="true" log="download-only" type="jar" />
+    <taskdef name="pmd" classname="net.sourceforge.pmd.ant.PMDTask" classpathref="pmd.classpath"/>
+  </target>
+
+  <target name="pmd" depends="pmd-typedef">
+    <sequential>
+      <echo message="PMD analyzing ${name}"/>
+      <mkdir dir="${common.build.dir}/pmd"/>
+      <pmd shortFilenames="true">
+	<!-- sourceLanguage name="java" version="1.8" / -->
+	<ruleset>${common.dir}/lucene-ruleset.xml</ruleset>
+        <formatter type="html" toFile="${common.build.dir}/pmd/${name}.html"/>
+        <fileset dir="${src.dir}">
+          <include name="**/*.java"/>
+          <exclude name="${pmd.excludes}"/>
+        </fileset>
+      </pmd>
+    </sequential>
+  </target>
+
+  <target name="lint" depends="pmd" />
+
   <!--+
       | M A C R O S
       +-->
diff --git lucene/lucene-ruleset.xml lucene/lucene-ruleset.xml
new file mode 100644
index 0000000..d1cab4e
--- /dev/null
+++ lucene/lucene-ruleset.xml
@@ -0,0 +1,46 @@
+<?xml version="1.0"?>
+<!--
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  -->
+<ruleset name="Lucene Ruleset" xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd"
+    xsi:noNamespaceSchemaLocation="http://pmd.sourceforge.net/ruleset_2_0_0.xsd">
+
+  <description>
+    Ruleset for Lucene/Solr
+  </description>
+
+  <rule ref="rulesets/java/strings.xml"/>
+  <rule ref="rulesets/java/unusedcode.xml"/>
+  <rule ref="rulesets/java/imports.xml"/>
+  <rule ref="rulesets/java/basic.xml">
+    <exclude name="EmptyCatchBlock" />
+    <exclude name="UselessOverridingMethod" />
+  </rule>
+
+  <!-- customize these rules -->
+  <rule ref="rulesets/java/empty.xml/EmptyCatchBlock"
+        message="Catch blocks must not be empty. They at least need a comment">
+    <priority>1</priority>
+    <properties>
+        <property name="allowCommentedBlocks" value="true" />
+    </properties>
+  </rule>
+  <rule ref="rulesets/java/unnecessary.xml/UselessOverridingMethod">
+    <priority>5</priority>
+  </rule>
+</ruleset>
diff --git solr/build.xml solr/build.xml
index d077946..f632ffc 100644
--- solr/build.xml
+++ solr/build.xml
@@ -198,6 +198,15 @@
       <propertyset refid="uptodate.and.compiled.properties"/>
     </ant>
   </target>
+
+  <target name="pmd">
+    <sequential>
+      <ant dir="core" target="pmd" inheritall="false"/>
+      <ant dir="solrj" target="pmd" inheritall="false"/>
+      <ant dir="test-framework" target="pmd" inheritall="false"/>
+      <contrib-crawl target="pmd" failonerror="true"/>
+    </sequential>
+  </target>
   
   <!-- Solrj targets -->
   <target name="test-solrj" description="Test java client">
-- 
1.9.1

